samcli/lib/providers/sam_function_provider.py

```diff
@@ -271,6 +294,17 @@ def _extract_functions(
 
                 # We don't care about other resource types. Just ignore them
 
+        # Apply filtering if function_logical_ids provided
+        if function_logical_ids:
+            function_logical_ids_set = set(function_logical_ids)
+            filtered_functions = {}
+
+            for full_path, function in result.items():
+                if SamFunctionProvider._should_include_function(function, function_logical_ids_set):
+                    filtered_functions[full_path] = function
```

Comment:
It's probably better to have this check inside the previous loop, so we don't need to create all the function resources that we're going to discard anyway.